contracts/test/integration/L2PlusSubnet.t.sol

```diff
@@ -0,0 +1,857 @@
+        executeTopDownMsgsAndExpectTopDownMessageEvent(
+            msgs,
+            params.subnet.gateway,
+            resultMessage,
+            params.subnetL3.subnetActorAddr,
+            params.subnet.gatewayAddr
+        );
+
+        // apply the result and check it was propagated to the correct actor
+        commitParentFinality(params.subnetL3.gatewayAddr);
+        executeTopDownMsgs(msgs, params.subnetL3.gateway);
+
+        assertTrue(params.caller.hasResult(), "missing result");
+        assertTrue(params.caller.result().outcome == params.expectedOutcome, "wrong result outcome");
+        assertTrue(keccak256(params.caller.result().ret) == keccak256(params.expectedRet), "wrong result outcome");
+    }
+
+    function sendCrossMessageFromParentToChildWithResult(Params memory params) public {
+        // register L2 into root network
+        registerSubnet(params.subnet.subnetActorAddr, params.root.gateway);
+        // register L3 into L2 subnet
+        registerSubnet(params.subnetL3.subnetActorAddr, params.subnet.gateway);
+
+        vm.deal(params.callerAddr, 1 ether);
+        vm.prank(params.callerAddr);
+
+        fundSubnet(params.root.gateway, params.subnet, params.callerAddr, 100000);
+
+        IpcEnvelope memory crossMessage = TestUtils.newXnetCallMsg(
+            IPCAddress({subnetId: params.root.id, rawAddress: FvmAddressHelper.from(params.callerAddr)}),
+            IPCAddress({subnetId: params.subnetL3.id, rawAddress: FvmAddressHelper.from(params.recipientAddr)}),
+            params.amount,
+            0
+        );
+
+        crossMessage.nonce = 1;
+        // send the cross message from the root network to the L3 subnet
+        vm.prank(params.callerAddr);
+        vm.expectEmit(true, true, true, true, params.root.gatewayAddr);
+        emit LibGateway.NewTopDownMessage({
+            subnet: params.subnet.subnetActorAddr,
+            message: crossMessage,
+            id: crossMessage.toHash()
+        });
+
+        crossMessage.nonce = 0;
+        params.root.gateway.messenger().sendContractXnetMessage{value: params.amount}(crossMessage);
+
+        IpcEnvelope[] memory msgs = new IpcEnvelope[](1);
+        msgs[0] = crossMessage;
+
+        // propagate the message from the L2 to the L3 subnet
+        executeTopDownMsgsAndExpectTopDownMessageEvent(
+            msgs,
+            params.subnet.gateway,
+            crossMessage,
+            params.subnetL3.subnetActorAddr,
+            params.subnet.gatewayAddr
+        );
+
+        // apply the cross message in the L3 subnet
+        executeTopDownMsgs(msgs, params.subnetL3.gateway);
+        assertEq(params.recipientAddr.balance, params.expectedAmount);
+
+        // submit checkoint so the result message can be propagated to L2
+        submitBottomUpCheckpoint(
+            callCreateBottomUpCheckpointFromChildSubnet(params.subnetL3.id, params.subnetL3.gateway),
+            params.subnetL3.subnetActor
```

Comment:
this should be `subnetL2.subnetActor`? The checkpoint should be propagated to L2 right? Not to L3 subnet actor.

Reply:
params.subnetL3.subnetActor corresponds to the L2 subnet actor. I agree it can be confusing, but I'm building around the already existing `TestSubnetDefinition`.

Reply:
I guess if you find it super confusing I can rename it. But it might affect a lot of files :)